external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/DirectKafkaInputDStream.scala

```diff
@@ -143,9 +147,14 @@ private[spark] class DirectKafkaInputDStream[K, V](
 
         lagPerPartition.map { case (tp, lag) =>
           val maxRateLimitPerPartition = ppc.maxRatePerPartition(tp)
-          val backpressureRate = Math.round(lag / totalLag.toFloat * rate)
-          tp -> (if (maxRateLimitPerPartition > 0) {
-            Math.min(backpressureRate, maxRateLimitPerPartition)} else backpressureRate)
+          val backpressureRate =
+            if (rate >= 0) {
+              val estimateRate = Math.round(lag / totalLag.toFloat * rate)
+              if (estimateRate > maxRateLimitPerPartition && maxRateLimitPerPartition > 0) {
+                maxRateLimitPerPartition
+              } else estimateRate
+            } else math.min(backpressureInitialRate, maxRateLimitPerPartition)
```

Comment:
I don't think these units are comparable, one is records / sec, and the other is records / sec / partition.